### ISSUE TITLE
Specifying in-line nested schemas

### ISSUE THREAD

**Issue by jonanderricchiuti** [fotinakis/swagger-blocks#42]
I have been trying to specify an in-line complex object which is the response of a get operation. This object is composed by other (nested) objects. Reading the swagger-blocks documentation I found a subsection (https://github.com/fotinakis/swagger-blocks#nested-complex-objects) that I think shows what I am trying to do.  Nevertheless I would like to maintain  the same specification standard to document my whole API.  I wrote the following swagger-blocks:

```
swagger_path mypath/resource do
    operation :get do
        .
        .
        .
        response 200 do
            key :description, 'Success!'
            schema do
                key :required, [:myObject]
                property :myobject  do
                    key :type, :object
                    schema do
                        key :required, [:nestedObject]
                        property :nestedObject
                            key :type, :object
                            schema do
                                key :required [:value]
                                property :value do
                                    key :type, :string
                                end
                            end
                        end
                    end
                end
            end
        end
    end
end
```

The problem with the above specification is that swagger-blocks does not allow to write a schema block inside of a property. I modified the PropertyNode class in the source code adding this:

```
def schema(inline_keys = nil, &block)                                      
        self.data[:schema] = Swagger::Blocks::SchemaNode.call(version: version, inline_keys: inline_keys, &block)
end
```

With that change I could accomplish what I wanted.

I would like to know if there is a better way to do what I did. If there is not,  what should I try instead?

**Comment by fotinakis** [comment]
I don't know if nesting a schema inside a schema is allowed in Swagger. According to the spec, it doesn't seem to be: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject -- though it may be deeper in the "JSON Schema Specification Draft 4" spec that I don't have time to dig through. ;)

Can you try getting what you want working in http://editor.swagger.io/? If you can, send me the YAML you end up with and I can see if we are not supporting that correctly in swagger-blocks.

**Comment by jonanderricchiuti** [comment]
I verified "JSON Schema Specification Draft 4" and I think that it's allowed. In the example section (http://json-schema.org/example1.html) under the name "Set of products schema:", I found an example that shows what I am trying to do. The example is as follows:

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "Product set",
    "type": "array",
    "items": {
      "title": "Product",
      "type": "object",
      "properties": {
        "id": {
          "description": "The unique identifier for a product",
          "type": "number"
        },
        "name": {
          "type": "string"
        },
        "price": {
          "type": "number",
          "minimum": 0,
          "exclusiveMinimum": true
        },
        "tags": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "minItems": 1,
          "uniqueItems": true
        },
        "dimensions": {
          "type": "object",
          "properties": {
            "length": {"type": "number"},
            "width": {"type": "number"},
            "height": {"type": "number"}
          },
          "required": ["length", "width", "height"]
        },
        "warehouseLocation": {
          "description": "Coordinates of the warehouse with the product",
          "$ref": "http://json-schema.org/geo"
        }
      },
      "required": ["id", "name", "price"]
    }
}
```
I wrote the following in http://editor.swagger.io/:

```json
{
  "swagger": "2.0",
    "info": {
      "version": "2.0.0",
      "title": "Test",
      "description": "Test",
      "contact": {
        "name": "Test Team"
      }
    },
    "tags": [{
      "name": "test",
      "description": "test"
    }],
    "host": "some.url.com",
    "basePath": "test",
    "consumes": ["application/json"],
    "produces": ["application/json"],
    "paths": {
      "swaggertest": {
        "get": {
          "tags": ["test"],
          "description": "test",
          "produces": ["application/json"],
          "responses": {
            "200": {
              "description": "Success!",
              "schema": {
                "required": ["myObject"],
                "properties": {
                  "myObject": {
                    "type": "object",
                    "schema": {
                      "required": ["nestedObject"],
                      "properties": {
                        "nestedObject": {
                          "type": "object",
                          "schema": {
                            "required": ["value"],
                            "properties": {
                              "value": {
                                "type": "string"
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
}
```

The page shows me a warning but when I open it, it does not show me any detail. Nevertheless in the response I got the schema that I wanted.
 
The following image shows what I explain above:
https://dl.dropboxusercontent.com/u/18749292/nestedswagger.png

**Comment by fotinakis** [comment]
Cool, that sounds right. We should allow `schema` inside `schema`. Feel like making a PR with your PropertyNode change above and adding a test?

**Comment by jonanderricchiuti** [comment]
I have never made a PR or a test in Ruby. If you could give me some guidance (perhaps what to read), I'd do it.

**Comment by fotinakis** [comment]
No problem, I'm happy to just do it or to help you through it.

But first, I think we may not be going down the right trail. I don't think `properties` are allowed to have `schema` inside of them in Swagger. I tried that in the Swagger Editor and it is an invalid response object.

But, I think the answer is this: don't use `schema`, just use more properties.

For example, this is VALID:

```
      responses:
        # Response code
        200:
          description: Successful response
          # A schema describing your response object.
          # Use JSON Schema format
          schema:
            title: ArrayOfPersons
            type: array
            items:
              title: Person
              type: object
              properties:
                name:
                  type: string
                single:
                  type: boolean
                nestedObject:
                  type: object
                  required:
                  - name
                  properties:
                    name:
                      type: string
```

But adding a `schema` makes it invalid:

```
              properties:
                name:
                  type: string
                single:
                  type: boolean
                nestedObject:
                  type: object
                  schema:
                    title: ArrayOfPersons
                    required:
                    - name
                    properties:
                      name:
                        type: string

```

So, can you try removing your nested `schema do` blocks and just doing the properties straight. I'm pretty sure that we support in swagger-blocks what you're trying to do, just that you haven't found the right swagger syntax yet. Make sure the Swagger Editor does what you want with zero errors, then move that format into swagger-blocks and see if it works.

If you find it's truly something we don't support in swagger-blocks, I can add support.

**Comment by jonanderricchiuti** [comment]
You are right, I should use `properties` instead `schema`. I'll try to move a format like the valid example above into swagger-blocks and I let you know the result.

**Comment by fotinakis** [comment]
Cool, feel free to re-open if so.

**Comment by jonanderricchiuti** [comment]
I tried to move the following valid example (I tested it in http://editor.swagger.io) to swagger-blocks:
```
responses:
        200:
          description: An array of products
          schema:
            title: ComplexObject
            type: object
            required: [name, nestedObject]
            properties: 
              name:
                type: string
              nestedObject:
                type: object
                required: [name, present, nestedObject]
                properties:
                  name:
                    type: string
                  present:
                    type: boolean
                  nestedObject:
                    type: object
                    properties:
                      value:
                        type: string
```
To achieve it I wrote the following code:
```ruby
require 'swagger/blocks'
class TestClass
  include Swagger::Blocks
  swagger_path '/pets/{id}' do
    operation :get do
      key :description, 'Returns a single pet if the user has access'
      key :operationId, 'findPetById'
      key :tags, [
        'pet'
      ]
      parameter do
        key :name, :id
        key :in, :path
        key :description, 'ID of pet to fetch'
        key :required, true
        key :type, :integer
        key :format, :int64
      end
      response 200 do
        key :description, 'pet response'
        schema do
          key :type, :object
          key :required, [:name, :nestedObject]
          property :name do
            key :type, :string
          end
          property :nestedObjet do
            key :type, :object
            key :required, [:name, :present, :nestedObject]
            property :name do
              key :type, :string
            end
            property :present do
              key :type, :boolean
            end
            property :nestedObjet do
              key :type, :object
              property :value do
                key :type, :string
              end
            end
          end
        end
      end
    end
  end
end
```

When I run the code above, I get:
```
swagger-blocks.rb:30:in `block (5 levels) in <class:TestClass>': undefined method `property' for #<Swagger::Blocks::PropertyNode:0x00000001d0dfa8> (NoMethodError)
```

I read the swagger-blocks code and inside `SchemaNode class`,  `property` is defined as:
```ruby
def property(name, inline_keys = nil, &block)
    self.data[:properties] ||= Swagger::Blocks::PropertiesNode.new
    self.data[:properties].version = version
    self.data[:properties].property(name, inline_keys, &block)
end
```
`PropertiesNode class`  as:
```ruby
class PropertiesNode < Node
    def property(name, inline_keys = nil, &block)
        self.data[name] = Swagger::Blocks::PropertyNode.call(version: version, inline_keys: inline_keys, &block)
    end
end
```
And  `PropertyNode class` does not define a `property`
```ruby
class PropertyNode < Node
    def items(inline_keys = nil, &block)
        self.data[:items] = Swagger::Blocks::ItemsNode.call(version: version, inline_keys: inline_keys, &block)
    end
end
```

**Comment by fotinakis** [comment]
Ok great, thanks for doing that. So the real issue here is that you can't nest `property` inside of `property` in swagger-blocks but it's supported by the spec. I will look at this now.

**Comment by fotinakis** [comment]
Done! https://github.com/fotinakis/swagger-blocks/commit/cf91f061d3b1370893fec28ef02000b20b0120a0 -- this has been released as gem v1.3.1.

For your reference, the way I did this was just updated our test JSON file, `swagger_v2_api_declaration.json`, with a `nestedObject` test data and then added the `nestedObject` property to `spec/lib/swagger_v2_blocks_spec.rb` (spec == test in Ruby land). Then I ran the tests with `bundle exec rspec`, saw that it failed with `undefined method `property' for #<Swagger::Blocks::PropertyNode>` just like you saw, then added the property method to `lib/swagger/blocks.rb`, ran the tests again and saw that they passed.
